src/cryoemservices/util/clem_array_functions.py

```diff
@@ -0,0 +1,873 @@
+"""
+Array manipulation functions to process and manipulate the images acquired via the Leica
+light microscope.
+"""
+
+from __future__ import annotations
+
+import itertools
+import logging
+import re
+from enum import Enum
+from pathlib import Path
+from typing import NamedTuple, Optional
+
+import numpy as np
+from tifffile import imwrite
+
+# Create logger object to output messages with
+logger = logging.getLogger("cryoemservices.util.clem_array_functions")
+
+
+"""
+HELPER CLASSES AND FUNCTIONS
+"""
+
+
+def get_valid_dtypes() -> tuple[str, ...]:
+    """
+    Use NumPy's in-built dtype dictionary to get list of available, valid dtypes.
+    Major dtype groups are "int", "uint", "float", and "complex"
+    """
+    # Extract list of NumPy dtype classes
+    dtype_class_list = list(
+        itertools.chain.from_iterable(
+            (
+                (str(value) for value in np.sctypes[key])
+                for key in np.sctypes.keys()
+                if key not in ("others",)
+            )
+        )
+    )
+    # Use regex matching to get just the dtype portion of the class
+    valid_dtypes: list[str] = []
+    pattern = r"<[a-z]+ '[a-z]+\.([a-z0-9]+)'>"
+    for dtype_class in dtype_class_list:
+        match = re.fullmatch(pattern, dtype_class)
+        if match is not None:
+            dtype = str(np.dtype(match.group(1)))
+            valid_dtypes.append(dtype)
+    if len(valid_dtypes) == 0:
+        logger.error("Unable to get list of NumPy dtypes from NumPy module")
+        raise Exception
+
+    return tuple(valid_dtypes)
+
+
+# Load valid dtypes for future use
+valid_dtypes = get_valid_dtypes()
+additional_dtype_keywords = (
+    "int",
+    "uint",
+    "float",
+    "complex",
+    "longdouble",
+    "clongdouble",
+)
+
+
+def get_dtype_info(dtype: str):
+    """
+    Returns NumPy's built-in dtype info object.
+
+    See the docs for:
+    numpy.finfo - https://numpy.org/doc/stable/reference/generated/numpy.finfo.html
+    numpy.iinfo = https://numpy.org/doc/stable/reference/generated/numpy.iinfo.html
+    """
+
+    # Validate input
+    if dtype not in valid_dtypes and not any(
+        # dtypes without numbers should also be accepted
+        dtype == key
+        for key in additional_dtype_keywords
+    ):
+        logger.error(f"{dtype} is not a valid NumPy dtype")
+        raise ValueError
+
+    # Load corresponding dictionary from NumPy
+    dtype_info = (
+        np.iinfo(dtype) if dtype.startswith(("int", "uint")) else np.finfo(dtype)
+    )
+    return dtype_info
+
+
+class Range(NamedTuple):
+    """
+    General named tuple containing range information.
+    """
+
+    min: int
+    max: int
+    range: int
+
+
+def get_dtype_limits(dtype: str):
```

Comment:
This function is fine, but might not be needed now as it just adds a range to the dtype_info